src/nativeScriptHashBuilder.c

```diff
@@ -0,0 +1,186 @@
+#include "nativeScriptHashBuilder.h"
+
+//#define TRACE_NATIVE_SCRIPT_HASH_BUILDER
+
+#ifdef TRACE_NATIVE_SCRIPT_HASH_BUILDER
+#define _TRACE(...) TRACE(__VA_ARGS__)
+#define _TRACE_BUFFER(...) TRACE_BUFFER(__VA_ARGS__)
+#else
+#define _TRACE(...)
+#define _TRACE_BUFFER(...)
+#endif // TRACE_NATIVE_SCRIPT_HASH_BUILDER
+
+#define APPEND_CBOR(type, value) blake2b_224_append_cbor_data(&builder->nativeScriptHash, type, value);
+#define APPEND_BUFFER(buffer, size) blake2b_224_append_buffer_data(&builder->nativeScriptHash, buffer, size);
+
+static void blake2b_224_append_buffer_data(
+        blake2b_224_context_t* hashCtx,
+        const uint8_t* buffer, size_t size
+)
+{
+	_TRACE_BUFFER(buffer, size);
+	blake2b_224_append(hashCtx, buffer, size);
+}
+
+__noinline_due_to_stack__
+static void blake2b_224_append_cbor_data(
+        blake2b_224_context_t* hashCtx,
+        uint8_t type, uint64_t value
+)
+{
+	uint8_t buffer[10];
+	size_t size = cbor_writeToken(type, value, buffer, SIZEOF(buffer));
+	_TRACE_BUFFER(buffer, size);
+	blake2b_224_append(hashCtx, buffer, size);
+}
+
+static inline void advanceState(native_script_hash_builder_t* builder)
+{
+	_TRACE("Advancing state, level = %u, remaining scripts = %u", builder->level, builder->remainingScripts[builder->level]);
+	// if the current number of remaining scripts is bigger than 0 we still
+	// expect scripts
+	if (builder->remainingScripts[builder->level] > 0) {
```

Comment:
I am not sure I understand. Currently the state has only two options in which it can be:
`NATIVE_SCRIPT_HASH_BUILDER_SCRIPT ` and `NATIVE_SCRIPT_HASH_BUILDER_FINISHED `. So in my opinion the switch statement you described doesn't make sense.
I added an assertion to check whether the state is `NATIVE_SCRIPT_HASH_BUILDER_SCRIPT `, because that's the only state this `advanceState` method should be called with. Is that okay with you?

Reply:
Generally, it's preferable to deal with enum values via a switch instead of a series of if/else statements: you can cleanly group several cases together, it is easy to modify (whether when the enum itself is modified or case handling is being modified) etc. So do not shy away from a switch even if there are just two values (usually, you also need `default` since the variable holding an enum value is not bound by its type to really only contain valid enum values).

What I wanted to say is that a function named `advanceState` suggests that we are going to look at the _current_ state and them move on to some kind of _following_ state. Your code of the function actually computed the new state from scratch without any regard to the current state. AFAIK all the advanceState/Stage functions work in the first way, not in the second. It might be OK to do it differently here, but we must have a clear reason for that.

Right now, I have noticed that `advanceState` is actually recursive? I would strongly prefer to avoid that if possible.

Reply:
I think there is a very clear reason to do it differently here. The state is determined only by the current count of remaining scripts for the given level. Basically the state always expects another script definition, be it the beginning of a complex script or a simple script. In my opinion there's nothing to be done to distinguish between expecting a complex vs a simple script as we don't know beforehand, so it is logical that both are represented by a single enum value (`NATIVE_SCRIPT_HASH_BUILDER_SCRIPT`).
The only state changes happens only when we received everything. That's when the count of remaining scripts for the 0th level is 0, then the state changes to second enum value (`NATIVE_SCRIPT_HASH_BUILDER_FINISHED`).
I don't see any logical way to incorporate other enum values, and thus no logical way to make this `advanceState` method work as the others in the codebase do.
Also since this method is only ever called, or should ever only be called, when the state is `NATIVE_SCRIPT_HASH_BUILDER_SCRIPT` I really don't see the point of having a switch statement here. If you were to argue that the default case of the switch could throw an error like invalid state then I think that's handled by the assertion at the top of the function.
```c
static inline void advanceState(native_script_hash_builder_t* builder)
{
	// advance state should be only called when current state processed a script
	ASSERT(builder->state == NATIVE_SCRIPT_HASH_BUILDER_SCRIPT);
```
So overall, maybe the "solution" is to name this function differently or maybe I can just add a comment that the state value is derived from the count of remaining scripts? (The second option is sufficient IMHO)

Reply:
An short explanatory comment would be sufficient. It is especially important to highlight the recursive nature of the function (I agree we want the recursion to walk back the levels), and what the recursion does (e.g. when it ends).

Reply:
I'd also suggest to split the if statement; it would markedly improve readability (at least for me), also some minor changes to comments:
```
	// if the current number of remaining scripts is bigger than 0 we still
	// expect scripts
	if (builder->remainingScripts[builder->level] > 0) {
		builder->state = NATIVE_SCRIPT_HASH_BUILDER_SCRIPT;
		return;
	}

	if (builder->level > 0) {
		// we are on some level bigger than 0 and expect the end of a complex script
		builder->level--;

		ASSERT(builder->remainingScripts[builder->level] > 0);
		builder->remainingScripts[builder->level]--;

		advanceState(builder);
	} else {
		// we know the whole native script is finished
		builder->state = NATIVE_SCRIPT_HASH_BUILDER_FINISHED;
	}
```

Reply:
I opted for a slightly different approach, I tried to make this function as simple as possible and added two more,
`simpleScriptFinished` and `complexScriptFinished`, analogous to the ones in the `deriveNativeScriptHash.c` file, which handle setting the correct `level` and `remainingScripts`.